### PR TITLE
FIX: DataObject::db returned fields in incorrect order, with incorrect data types

### LINKED ISSUE #3802: Form Fields scaffolded from $db are now reversed in order
After changes released in 3.1.9, (see #3620) DataObject->db() now reverses the order of $db fields so child DataObjects can override their parents $db fields. But because FormScaffolder->getFieldList() uses DataObject->db() to scaffold form fields, we're getting scaffolded form fields in reverse order too.

### PR DESCRIPTION
**Pull request by kinglozzer**
fixes #3802

### CODE REVIEW

**Comment by kinglozzer** [comment]
@tractorcow or as an alternative, refactored into two methods as you suggested: https://github.com/kinglozzer/sapphire/commit/ed07bfadb3ddf53c118a3a65bb11db67cc8091a8

**Comment by tractorcow** [comment]
For 3.1 this is going to be fine. :)

@sheadawson  can you check this fixes your issue please?

File: model/DataObject.php [hunk 1]
```diff
@@ -1666,9 +1666,14 @@ public function belongs_to($component = null, $classOnly = true) {
...
-		$items = array();
 
-		foreach(array_reverse($classes) as $class) {
+		// If we're looking for a specific field, we want to hit subclasses first as they may override field types
+		if($fieldName) {
+			$classes = array_reverse($classes);
+		}
+
+		$items = array();
+		foreach($classes as $class) {
```
Comment (tractorcow) [inline model/DataObject.php]: Don't forget to re-instate the exception for `$class === 'DataObject'`

**Comment by tractorcow** [comment]
Thanks @kinglozzer really appreciate your prompt response in this. :)

**Comment by sheadawson** [comment]
Yeap fixes my issue thanks guys!

**Comment by tractorcow** [comment]
Thanks both @sheadawson and @kinglozzer

**Pull request merged**